Add textDocument/inlayHint and experimental/inlayHints

topisani opened the pull request:
Close #600 

For now this doesnt remove the old rust-analyzer api, and adds both the `textDocument/inlayHint` and `experimental/inlayHints`. The former is proposed for LSP v3.17, and the latter is the temporary name used for the proposed api by rust-analyzer.

I think we should remove both old apis when the new one is standardized, and adapted by rust-analyzer, since i have seen no indication that other servers have used the rust-analyzer or experimental prefix versions.

krobelus commented on rc/lsp.kak:
```diff
@@ -1158,6 +1158,39 @@ incomingOrOutgoing = $1
 }
 
+
+define-command lsp-inlay-hints -docstring "lsp-inlay-hints: Request inlay hints (lsp)" %{
```
I think we should drop the `(lsp)` suffix because it will become obsolete once we remove the rust-analyzer variant.
I also prefer lowercase after colon (like most Kakoune commands) but that's an unrelated change.

```suggestion
define-command lsp-inlay-hints -docstring "lsp-inlay-hints: request inlay hints" %{
```

krobelus commented on src/language_features/inlay_hints.rs:
```diff
@@ -0,0 +1,105 @@
+};
+
+#[derive(Debug, PartialEq, Clone, Default, Serialize, Deserialize)]
+struct InlayHintsOptions {
```
good point defining this here. I just realized that a lot of types in `src/types.rs` should rather be in the module where they're used.
The name collision with InlayHintsParams is ugly but fair enough.

krobelus commented on src/language_features/inlay_hints.rs:
```diff
@@ -0,0 +1,105 @@
+                ))
+            },
+        )
+        .collect::<Vec<String>>()
```
unrelated but there's no need to collect() before join() (if we import `itertools::Itertools`)

krobelus commented on src/util.rs:
```diff
@@ -52,12 +52,12 @@ impl Drop for TempFifo {
 /// Escape Kakoune string wrapped into single quote
 pub fn editor_escape(s: &str) -> String {
-    s.replace("'", "''")
+    s.replace('\'', "''")
```
not sure about this. If we want this change it should be in a separate commit, and handle other single-char replace like in `src/language_features/hover.rs` too.

topisani replied:
Oh yeah, I just naively resolved the clippy fixmes out of habit - I'll clear them out of this

krobelus replied:
Huh, I didn't see clippy complaining. OTOH I do get code actions from rust-analyzer to transform string -> char and vice versa, but those are just assists.

topisani replied:
It might have been those - i may have just assumed they were from clippy

topisani replied:
https://rust-lang.github.io/rust-clippy/master/index.html#single_char_pattern
Ah, yes, there it is

krobelus replied:
strange that my clippy doesn't suggest that. Anyway, since it's blessed by clippy the idea seems okay.
If possible we should have a reproducible `cargo clippy` config, not sure how that would work

krobelus commented:
looks nice, I have never used inlay hints before; they seem much less invasive than inlay diagnostics

topisani commented on src/language_features/inlay_hints.rs:
```diff
@@ -0,0 +1,105 @@
+                    ""
+                };
+                // Escape markup inside label
+                let label = escape_tuple_element(&label).replace('{', "\\{");
```
The doubly escaped backslashes are why I inlined the markup escape - this works, your suggestion does not, since it doubly escapes the backslashes, undoing the markup escaping

krobelus replied:
wow I had no idea. This means that if the label contains \{} then there is no way to render it properly; we are either losing the backslash or the braces.
The markup seems more important, so LGTM

krobelus replied:
I can't reproduce the problem with my suggestion anymore.
OTOH, the current version fails to display inlay hints if I add something like `let label = r#"123\{456"#;` above.

This command illustrates that it works fine to double backslashes at every level of escaping (in addition to escaping { and |)

```
kak -e '
	exec %{%dihello world<esc>}
	declare-option range-specs my_ranges
	add-highlighter global/ replace-ranges my_ranges
	# raw value:                      \{
	#    -> escaped for markup:     \\\{
	#      -> escaped for tuple: \\\\\\{
	set-option global my_ranges %val{timestamp} %<1.5+0|\\\\\\{>
	# renders as hell\{o world
'
```

will commit a fix

topisani commented:
This should address the comments - i added `lsp-inlay-hints-{enable,disable}`, and I added docs and an entry in the changelog

As mentioned, i havent removed the old rust-analyzer-inlay-hints api, but i guess we could do that as well?

krobelus commented:
Nice changes, thanks! They should be squashed before merging so the history is less noisy. I generally squash instead of pushing fixup commits.

I think instead of removing `rust-analyzer-inlay-hints` we should just make it delegate to `lsp-experimental-inlay-hints`.
(and as soon as rust-analyzer switches we can change it to `lsp-inlay-hints`).
This avoids breaking existing configurations, at essentially no cost (save for confusing users who are switching back to old rust-analyzer versions and expect that `rust-analyzer-inlay-hints` would do the old thing). We'll still deprecate the command (removing it from docs and making it hidden).

krobelus commented on rc/lsp.kak:
```diff
@@ -1873,6 +1906,31 @@ define-command lsp-stop-on-exit-disable -docstring "Don't end kak-lsp session on
 
+define-command lsp-inlay-hints-enable -params 1 -docstring "lsp-inlay-hints-enable <scope>: Enable inlay hints for <scope>" %{
+    add-highlighter "%arg{1}/lsp_inlay_hints" replace-ranges lsp_inlay_hints
+    hook -group lsp-inlay-hints %arg{1} BufReload .* lsp-inlay-hints
```
I guess this means we should make `lsp-inlay-hints` hidden because most users will be happy with `lsp-inlay-hints-enable`

topisani commented:
Alright, i removed the rust-analyzer/inlayHints implementation, and let the kak commands delegate to lsp-experimental-inlay-hints.

I rebased and squashed, so this should be ready for merge
